hbase-server/src/test/java/org/apache/hadoop/hbase/replication/regionserver/TestReplicationSourceWithoutReplicationZnodes.java

```diff
@@ -0,0 +1,234 @@
+    conf.setInt("replication.source.maxretriesmultiplier", 10);
+    utility = new HBaseTestingUtility(conf);
+    utility.startMiniZKCluster();
+
+    zkw = new ZooKeeperWatcher(conf, "test", null);
+    ZKUtil.createWithParents(zkw, "/hbase/replication");
+    ZKUtil.createWithParents(zkw, "/hbase/replication/peers/1");
+    ZKUtil.setData(zkw, "/hbase/replication/peers/1",
+      Bytes.toBytes(conf.get(HConstants.ZOOKEEPER_QUORUM) + ":"
+        + conf.get(HConstants.ZOOKEEPER_CLIENT_PORT) + ":/1"));
+    ZKUtil.createWithParents(zkw, "/hbase/replication/peers/1/peer-state");
+    ZKUtil.setData(zkw, "/hbase/replication/peers/1/peer-state",
+      ReplicationStateZKBase.ENABLED_ZNODE_BYTES);
+    ZKUtil.createWithParents(zkw, "/hbase/replication/state");
+    ZKUtil.setData(zkw, "/hbase/replication/state", ReplicationStateZKBase.ENABLED_ZNODE_BYTES);
+
+    ZKClusterId.setClusterId(zkw, new ClusterId());
+    FSUtils.setRootDir(utility.getConfiguration(), utility.getDataTestDir());
+    fs = FileSystem.get(conf);
+    oldLogDir = new Path(utility.getDataTestDir(),
+      HConstants.HREGION_OLDLOGDIR_NAME);
+    logDir = new Path(utility.getDataTestDir(),
+      HConstants.HREGION_LOGDIR_NAME);
+    server = new DummyServer(conf, "example.hostname.com", zkw);
+    replication = new Replication(server, fs, logDir, oldLogDir);
+    manager = replication.getReplicationManager();
+
+    manager.addSource(slaveId);
+
+    htd = new HTableDescriptor(test);
+    HColumnDescriptor col = new HColumnDescriptor(f1);
+    col.setScope(HConstants.REPLICATION_SCOPE_GLOBAL);
+    htd.addFamily(col);
+    col = new HColumnDescriptor(f2);
+    col.setScope(HConstants.REPLICATION_SCOPE_LOCAL);
+    htd.addFamily(col);
+
+    hri = new HRegionInfo(htd.getTableName(), r1, r2);
+  }
+
+  @AfterClass
+  public static void tearDownAfterClass() throws Exception {
+    try {
+      manager.join();
+    } catch (RuntimeException re) {
+      if (re.getMessage().equals(fakeExceptionMessage)) {
+        LOG.info("It is fine");
+      }
+    }
+    utility.shutdownMiniCluster();
+  }
+
+  @Rule
+  public TestName testName = new TestName();
+
+  private void cleanLogDir() throws IOException {
+    fs.delete(logDir, true);
+    fs.delete(oldLogDir, true);
+  }
+
+  @Before
+  public void setUp() throws Exception {
+    LOG.info("Start " + testName.getMethodName());
+    cleanLogDir();
+  }
+
+  @After
+  public void tearDown() throws Exception {
+    LOG.info("End " + testName.getMethodName());
+    cleanLogDir();
+  }
+
+  /**
+   * When the peer is removed, hbase remove the peer znodes and there is zk watcher
+   * which terminates the replication sources. In case of zk watcher not getting invoked
+   * or a race condition between source deleting the log znode and zk watcher
+   * terminating the source, we might get the NoNode exception. In that case, the right
+   * thing is to terminate the replication source.
```

Comment:
>The only place I see updating log position (eventually calling ReplicationQueuesZKImpl.removeLog) is the SourceShipperThread, which is terminated in ReplicationSouceManager.removePeer.

You are right, but the `ReplicationSouceManager.removePeer` is invoked by a `PeersWatcher` zk listerner.
So, this means there is a race condition between `ReplicationQueuesZKImpl.removeLog` and `ReplicationSouceManager.removePeer` which terminates the source, even though race condition won't impact multiple region servers at once. 
But if the zk listerner is not invoked at all due to some zk issue, this will effect all the region servers which are replicating and will badly affect the availability. We have seen this issue recently and it seems like ZK issue. But handling the NoNode exception gracefully will make RS resilient to this issue.

Reply:
These are the exceptions for reference:


```bash

2021-02-16 20:11:58,567 FATAL [95922885,xyz_peer] regionserver.HRegionServer - ABORTING region server regionserver-111,60020,1613495922885: Failed to remove wal from queue (queueId=xyz_peer, filename=regionserver-111%2C60020%2C1613495922885.1613505863058)
org.apache.zookeeper.KeeperException$NoNodeException: KeeperErrorCode = NoNode for /hbase/replication/rs/regionserver-111,60020,1613495922885/xyz_peer/regionserver-111%2C60020%2C1613495922885.16135058630
58
        at org.apache.zookeeper.KeeperException.create(KeeperException.java:114)
        at org.apache.zookeeper.KeeperException.create(KeeperException.java:54)
        at org.apache.zookeeper.ZooKeeper.delete(ZooKeeper.java:890)
        at org.apache.hadoop.hbase.zookeeper.RecoverableZooKeeper.delete(RecoverableZooKeeper.java:238)
        at org.apache.hadoop.hbase.zookeeper.ZKUtil.deleteNode(ZKUtil.java:1341)
        at org.apache.hadoop.hbase.zookeeper.ZKUtil.deleteNode(ZKUtil.java:1330)
        at org.apache.hadoop.hbase.replication.ReplicationQueuesZKImpl.removeLog(ReplicationQueuesZKImpl.java:142)
        at org.apache.hadoop.hbase.replication.regionserver.ReplicationSourceManager.cleanOldLogs(ReplicationSourceMana
ger.java:232)
        at org.apache.hadoop.hbase.replication.regionserver.ReplicationSourceManager.cleanOldLogs(ReplicationSourceMana
ger.java:222)
        at org.apache.hadoop.hbase.replication.regionserver.ReplicationSourceManager.logPositionAndCleanOldLogs(Replica
tionSourceManager.java:198)
        at org.apache.hadoop.hbase.replication.regionserver.ReplicationSource$ReplicationSourceShipperThread.updateLogP
osition(ReplicationSource.java:831)
        at org.apache.hadoop.hbase.replication.regionserver.ReplicationSource$ReplicationSourceShipperThread.shipEdits(
ReplicationSource.java:746)
        at org.apache.hadoop.hbase.replication.regionserver.ReplicationSource$ReplicationSourceShipperThread.run(Replic
ationSource.java:650)
```

Reply:
@wchevreuil Let me know if that answers your concern or you have any other feedback on this.

Reply:
Ok, so the problem can happen because `ReplicationSouceManager.removePeer` is only invoked after the peer node already got deleted in ZK, whilst the `SourceShipperThread` might be trying update the znode in between. Thanks for explaining it further, @sandeepvinayak !